File: Problemset/zi-fu-chuan-de-pai-lie-lcof/zi-fu-chuan-de-pai-lie-lcof.py

# @Title: 字符串的排列 (字符串的排列  LCOF)
# @Author: cocofe
# @Date: 2020-08-16 15:11:36
# @Runtime: 100 ms
# @Memory: 18.9 MB

class Solution(object):
    def permutation(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        
        chars, ans, length = list(s), [], len(s)
        # idx 之前是固定住的, 准备固定idx
        def dfs(idx):
            if idx == length - 1:
                ans.append(''.join(chars))
                return
            to_fixed_chars = set()
            for i in range(idx, length):
                if chars[i] in to_fixed_chars:
                    continue
                to_fixed_chars.add(chars[i])
                # 选择一个字符, 固定在idx位置, 并原有的idx字符进行交换
                chars[i], chars[idx] = chars[idx], chars[i]
                # 固定idx+1位置
                dfs(idx + 1)
                # 恢复chars, 继续选择下一个字符固定在idx位置
                chars[i], chars[idx] = chars[idx], chars[i]
        dfs(0)
        return ans

File: Problemset/zi-fu-chuan-de-pai-lie-lcof/zi-fu-chuan-de-pai-lie-lcof.java

// @Title: 字符串的排列 (字符串的排列  LCOF)
// @Author: cocofe
// @Date: 2021-06-22 02:05:10
// @Runtime: 17 ms
// @Memory: 43.3 MB

class Solution {
    List<String> ans = new LinkedList<>();
    public String[] permutation(String s) {
    char[] chs = s.toCharArray();
    Arrays.sort(chs);
    backtrace(0, chs, new StringBuilder());
    String[] arrs = new String[ans.size()];
    return ans.toArray(arrs);
    }

    public void swap(char[] chs, int pos1, int pos2){
        if (pos1 == pos2) {
            return;
        }
        char tmp = chs[pos1];
        chs[pos1] = chs[pos2];
        chs[pos2] = tmp;
    }

    public void backtrace(int pos, char[] chs, StringBuilder chars){
        if(pos >= chs.length){
            ans.add(chars.toString());
            return;
        }
        HashSet fixed = new HashSet();
        for (int i=pos; i<chs.length; i++) {
            char ch = chs[i];
            if(fixed.contains(ch)){
                continue;
            }
            fixed.add(ch);
            chars.append(ch);
            swap(chs, pos, i);
            backtrace(pos+1, chs, chars);
            swap(chs, pos, i);
            chars.deleteCharAt(chars.length() - 1);
        }
    }
    
}
